suppress/suppress.go

```diff
@@ -43,45 +39,47 @@ func For(duration time.Duration, id string, f func(int, string)) {
 // this function should have depth >= 1.
 func WrapFor(depth int, duration time.Duration, id string, f func(int, string)) {
 	pc, file, line, _ := runtime.Caller(depth)
-	file = filepath.Base(file)
-	key := fmt.Sprintf("%d%s", pc, id)
-	tag := fmt.Sprintf("%s:%d %s", file, line, id)
+	key := locKey{pc, id}
 
-	_suppressor.Lock()
-	state, exists := _suppressor.states[key]
+	lock.RLock()
+	state, exists := suppressors[key]
+	lock.RUnlock()
 	if !exists {
-		state = new(suppressorState)
-		_suppressor.states[key] = state
+		lock.Lock()
+		state, exists = suppressors[key]
+		if !exists {
+			state = &suppressorState{count: -1}
+			suppressors[key] = state
+		}
+		lock.Unlock()
 	}
 
-	state.count++
-	state.lastFunc = f
+	if atomic.AddInt64(&state.count, 1) > 0 {
+		return
+	}
 
-	if state.count == 1 {
-		now := time.Now()
-		if state.next.Before(now) {
-			// no need to suppress
-			state.lastFunc(state.count, tag)
-			state.next = now.Add(duration)
-			state.count = 0
-		} else {
-			// this is the first time we've suppressed, so schedule the next
-			// firing. subsequent passes will only increment the counter.
-			go func() {
-				_suppressor.Lock()
-				wait := state.next.Sub(time.Now())
-				_suppressor.Unlock()
+	tag := fmt.Sprintf("%s:%d %s", filepath.Base(file), line, id)
 
-				time.Sleep(wait)
+	// Subsequent calls will be suppressed while this goroutine is running.
+	// Every `duration`, it checks to see how many calls have been suppressed.
+	// If there have been none, it exits, and sets the suppressorState count to
+	// -1, indicating that there is no active suppression.
+	go func() {
+		ticker := time.NewTicker(duration)
+		defer ticker.Stop()
 
-				_suppressor.Lock()
-				state.lastFunc(state.count, tag)
-				state.next = time.Now().Add(duration)
-				state.count = 0
-				_suppressor.Unlock()
-			}()
+		for range ticker.C {
+			count := atomic.LoadInt64(&state.count)
+			if count == 0 {
```

Comment:
it doesn't need to remove the key from the map, we set count to -1.   That means the next call will trigger it again here: https://github.com/fastly/go-utils/pull/29/files#diff-655af7c92fb732b43708efa830bcda14R57